primitives/runtime/src/traits.rs

```diff
@@ -104,6 +119,12 @@ impl Verify for sp_core::sr25519::Signature {
 	}
 }
 
+impl BatchVerify for sp_core::sr25519::Signature {
+	fn batch_verify<L: Lazy<[u8]>>(&self, mut msg: L, signer: &sp_core::sr25519::Public) -> bool {
+		sp_io::crypto::sr25519_batch_verify(self, msg.get(), signer)
+	}
+}
+
 impl Verify for sp_core::ecdsa::Signature {
```

Comment:
Why not for `ecdsa`?

Reply:
ecdsa is a bit tricky, since it dominates the verification (takes much more time compared to transaction execution)

Node can be forced to do much more work before first ecdsa is resolved as invalid

Reply:
The other signatures probably also take much less time to verify. This feature is ridiculous if you draw the line here. Than we don't need async verification at all. If this would be a problem in the future, we could disable async signature verification when importing blocks on the tip of the chain. However, for syncing we want the best performance.

Reply:
To me, you can trick node to verify 1000 or so heavy signatures for free until first one is resolved and possibly completely stop some validators, not going to add it here until some proof that this is safe provided.

Reply:
I can also force you to verify 3000 sr25519 signatures. Where it the difference?

Reply:
How?

Reply:
Also, 3000 sr25519 is ~50ms if batched
while 1000 edsca time is measured in seconds

Reply:
We've the ecRecover variant of ECDSA that makes batch verification impossible.

Reply:
We only batch 128 or whatever sr25519 signatures, so not 3k in one batch.

Secondly, we don't use batching for ed25519, so the same applies as for ecdsa. Maybe we should just name this stuff async verification and not force us here to go with batching, because only one signature is batched in random sized batches.

Reply:
If you really want to prevent other nodes spamming us with blocks of a lot of invalid signatures, we need to improve the import. We should return a specific error on an invalid transaction signature and ban this node immediately. 

And in the moment we find the first invalid signature, we abort the import anyway. Finding the first invalid signature would probably not take that many transactions, because the transaction itself also takes some time. And you can play the same attack for < 127 sr25519 signatures. However, a heavy transaction normally also requires that the account has some funds and if you don't have them, the transaction is aborted directly anyway.

Reply:
That is the point, you can schedule much more verifications while edsca resolves (which resolves orders of magnitude slower than sr25519/ed25519, batched or not)

Reply:
Okay, you basically just want early exit of the scheduled verifications? So that we abort all, if one fails? Should not be that hard to implement, just make the future select on an exit signal.

And ECDSA is also not orders of magnitudes slower, on my machine it was factor 3 slower.

Reply:
We can support Ed25519 batch verification eventually I think, but not worth the time right now.

Reply:
Ok ECDSA performance is better than I had impression of, indeed just 3x thanks @bkchr, https://github.com/paritytech/substrate/pull/6697 

So should be no problem adding it also

Reply:
Nevertheless, I like the idea of adding an early bail out for these verification tasks if one failed. This should prevent the problems you have described here.

Reply:
I added implementation for standalone ECDSA, but I think the way it works it can't be used in `MultiSignature`, since it uses `AccountId32`, which can store only `blake2(public)`, but not public itself